### PR TITLE
[ONNX] Added translator for MatMulNBits from com.microsoft domain

### PR DESCRIPTION
**Pull request by gkrivor**
### Details:
 - Added translator for MatMulNBits from com.microsoft domain

### Tickets:
 - CVS-152263

### CODE REVIEW

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (rkazants) [inline src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:219]: can you try to use `transpose_a` and `transpose_b` attributes of `MatMul` to eliminate Transpose nodes in this decompositions?

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (rkazants) [inline src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:99]: Should we really check it? Can we assume that they are the same by specification in correct model?

Comment (rkazants) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:99]: also, not good because `get_element_type()` can return dynamic or undefined type for some tensors.

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (rkazants) [inline src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:39]:
```suggestion
    ov::Output<ov::Node> bias;                                                           // optional, input[5]
```

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:99]: Docs says they should have same types. Yes, I know about possible issue with dynamic type. But currently it fails at a constant creation step - I cannot create "dynamic" constant for the default_zp. I'm trying remember, did you do something similar like "ConstantLike" in TF?

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:219]: Let me try move this return to another place... I'll get back...

Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:219]: Looks like it is possible to remove at least one Transponse operation in case no bias provided for n_blocks_per_col>1.

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (slyalin) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:146]: Could you put more description in the comments to help follow the blocking schema and what we are trying to achieve with the decomposition?

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (slyalin) [inline src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:158]: This transpose runs in low bit precision: `u2`/`u4`. It may have the same issues as `Slice`. Have you checked the correctness of this particular step?

Then I'm not quite sure what this `Transpose` gives in terms of the original matrix B. It looks to permute also the blocked dimension, but due to that multiplier that we applied when creating `converted_b` depending on the type of elements. Could you provide some visuals or extended description, or link where it is described in more detailed form in comparison to the original ORT spec?

**Review by slyalin — CHANGES_REQUESTED** [review]

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:141]: No, need to make a reproducer and other steps... I'm goning to do it this.

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:118]: Operations below (sub, multiple) operates with tensors [N * n_blocks_per_col]. Do you mean it is wrong reshaping approach?

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:146]: B contains a quantized data blocks. Block size should be power of 2 and not less than 16. In case it isn't aligned - rest of a block is filled by zeros. For example, if data is (uint8)[1,2,3,4,5,6] then block will be (uint8)[1,2,3,4,5,6,0,0,0,0,0,0,0,0,0,0]. And last zeros are unused.

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (andrei-kochin) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:42]: Actually as we are targeting the functional enabling first discussion of this topic with plugin teams is optional

**Review by rkazants — APPROVED** [review]

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (rkazants) [inline src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:142]: please clarify what TODO is required (some message) and ticket

File: src/frontends/onnx/tests/models/com.microsoft/matmulnbits_3x4.onnx
Comment (gkrivor) [reply src/frontends/onnx/tests/models/com.microsoft/matmulnbits_3x4.onnx:1]: They are very small and non-editable. Is it necessary?
I still want to make a regular *.prototxt for them before merge.

**Review by andrei-kochin — APPROVED** [review]

File: src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp:142]: Have to prepare a reproducer before

File: src/frontends/onnx/tests/models/com.microsoft/matmulnbits_3x4.onnx
Comment (gkrivor) [reply src/frontends/onnx/tests/models/com.microsoft/matmulnbits_3x4.onnx:1]: Replaced by *.prototxt

**Review by slyalin — APPROVED** [review]
If it is *functional* enabling only that will unblock further development, then I agree to merge.

**Pull request merged**
